src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,804 @@
+	const char *runtime_name = NULL;
+	const char *json_url = NULL;
+	guint spec = 0;
+
+	parser = json_parser_new ();
+
+	ret = json_parser_load_from_data (parser, data, -1, error);
+	if (!ret)
+		return NULL;
+
+	root = json_node_get_object (json_parser_get_root (parser));
+	if (!root) {
+		g_set_error (error,
+		             GS_PLUGIN_ERROR,
+		             GS_PLUGIN_ERROR_FAILED,
+		             "no root object");
+		return NULL;
+	}
+
+	spec = json_object_get_int_member (root, JSON_SPEC_KEY);
+	if (spec != EXTERNAL_ASSETS_SPEC_VERSION) {
+		g_set_error (error,
+		             GS_PLUGIN_ERROR,
+		             GS_PLUGIN_ERROR_FAILED,
+		             "External asset's json spec version '%u' does "
+			     "not match the plugin. Expected '%u'", spec,
+			     EXTERNAL_ASSETS_SPEC_VERSION);
+		return NULL;
+	}
+
+	runtime = json_object_get_object_member (root, JSON_RUNTIME_KEY);
+	if (!runtime) {
+		g_set_error (error,
+		             GS_PLUGIN_ERROR,
+		             GS_PLUGIN_ERROR_FAILED,
+		             "External asset's json has no '%s' member set",
+			     JSON_RUNTIME_KEY);
+		return NULL;
+	}
+
+	runtime_name = json_object_get_string_member (runtime,
+						      JSON_RUNTIME_NAME_KEY);
+	if (!runtime_name) {
+		g_set_error (error,
+		             GS_PLUGIN_ERROR,
+		             GS_PLUGIN_ERROR_FAILED,
+		             "External asset's runtime member has no '%s' key "
+			     "set", JSON_RUNTIME_NAME_KEY);
+		return NULL;
+	}
+
+	json_url = json_object_get_string_member (runtime, JSON_RUNTIME_URL_KEY);
+	if (!json_url) {
+		g_set_error (error,
+		             GS_PLUGIN_ERROR,
+		             GS_PLUGIN_ERROR_FAILED,
+		             "External asset's runtime member has no '%s' key "
+			     "set", JSON_RUNTIME_URL_KEY);
+		return NULL;
+	}
+
+	*url = g_strdup (json_url);
+
+	return g_strdup (runtime_name);
+}
+
+static GsApp *
+gs_plugin_get_app_external_runtime (GsPlugin *plugin,
+				    GsApp *headless_app)
+{
+	GsApp *runtime = NULL;
+	GsPluginData *priv;
+	g_autofree char *id = NULL;
+	g_autofree char *full_id = NULL;
+	g_autofree char *url = NULL;
+	g_autofree char *json_data = NULL;
+	g_autoptr (GError) error = NULL;
+	const char *metadata;
+
+	metadata = gs_app_get_metadata_item (headless_app,
+					     METADATA_EXTERNAL_ASSETS);
+
+	if (!metadata)
+		return NULL;
+
+	json_data = g_uri_unescape_string (metadata, NULL);
+	id = extract_runtime_info_from_json_data (json_data, &url, &error);
+
+	if (!id) {
+		g_debug ("Error getting external runtime from "
+			 "metadata: %s", error->message);
+		return NULL;
+	}
+
+	priv = gs_plugin_get_data (plugin);
+
+	full_id = g_strdup_printf ("%s:%s",
+				   gs_flatpak_get_prefix (priv->flatpak),
+				   id);
+
+	runtime = gs_plugin_cache_lookup (plugin, full_id);
+	if (runtime) {
+		g_debug ("Found cached '%s'", full_id);
+		/* Adopt runtime if needed */
+		if (g_strcmp0 (gs_app_get_management_plugin (runtime),
+			       gs_plugin_get_name (plugin)) != 0) {
```

Comment:
This is because *gs_app_set_management_plugin* always frees+reassigns the string. I will instead move the logic to GsApp and ask Richard if not having it was intentional somehow.